File: src/main/java/com/telegram/codex/conversation/application/job/JobSchedulerService.java
package com.telegram.codex.conversation.application.job;

import com.telegram.codex.conversation.application.reply.ReplyGenerationService;
import com.telegram.codex.conversation.application.session.SessionService;
import com.telegram.codex.conversation.infrastructure.MediaGroupBufferRepository;
import com.telegram.codex.integration.telegram.application.CompactResultSender;
import com.telegram.codex.integration.telegram.application.webhook.InboundMessageProcessor;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import jakarta.annotation.PreDestroy;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

@Service
public class JobSchedulerService {

    private final MediaGroupBufferRepository mediaGroupStore;
    private final ObjectProvider<InboundMessageProcessor> inboundMessageProcessorProvider;
    private final ReplyGenerationService replyGenerationService;
    private final ScheduledExecutorService scheduledExecutorService;
    private final SessionService sessionService;
    private final CompactResultSender compactResultSender;
    private final ExecutorService taskExecutor;

    public JobSchedulerService(
        MediaGroupBufferRepository mediaGroupStore,
        ObjectProvider<InboundMessageProcessor> inboundMessageProcessorProvider,
        ReplyGenerationService replyGenerationService,
        SessionService sessionService,
        CompactResultSender compactResultSender
    ) {
        this.mediaGroupStore = mediaGroupStore;
        this.inboundMessageProcessorProvider = inboundMessageProcessorProvider;
        this.replyGenerationService = replyGenerationService;
        this.sessionService = sessionService;
        this.compactResultSender = compactResultSender;
        this.taskExecutor = Executors.newVirtualThreadPerTaskExecutor();
        this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(Thread.ofPlatform()
            .name("media-group-scheduler-", 0)
            .daemon(true)
            .factory());
    }

    public void enqueueReplyGeneration(InboundMessage message) {
        taskExecutor.execute(() -> replyGenerationService.handle(message));
    }

    public void scheduleMediaGroupFlush(String key, long expectedDeadlineAt, Duration waitDuration) {
        scheduledExecutorService.schedule(
            () -> taskExecutor.execute(() -> flushMediaGroup(key, expectedDeadlineAt)),
            waitDuration.toMillis(),
            TimeUnit.MILLISECONDS
        );
    }

    public void enqueueSessionCompact(String chatId) {
        taskExecutor.execute(() -> {
            var result = sessionService.compact(chatId);
            compactResultSender.send(chatId, result);
        });
    }

    @PreDestroy
    public void shutdown() {
        scheduledExecutorService.shutdown();
        taskExecutor.shutdown();
    }

    private void flushMediaGroup(String key, long expectedDeadlineAt) {
        MediaGroupBufferRepository.FlushResult result = mediaGroupStore.flush(key, expectedDeadlineAt);
        switch (result.status()) {
            case "ready" -> inboundMessageProcessorProvider.getObject().process(result.message());
            case "pending" -> scheduleMediaGroupFlush(
                key,
                expectedDeadlineAt,
                Duration.ofMillis(Math.round(result.waitDurationSeconds() * 1000.0))
            );
            default -> {
            }
        }
    }
}

File: src/conversation/job-scheduler.service.ts
import {forwardRef, Inject, Injectable, Logger, OnModuleDestroy} from "@nestjs/common";
import {InboundMessage} from "../telegram/inbound-message";
import {CompactResultSenderService} from "../telegram/compact-result-sender.service";
import {MediaGroupBufferRepository} from "./media-group-buffer.repository";
import {ReplyGenerationService} from "./reply-generation.service";
import {SessionService} from "./session.service";

export interface InboundMessageProcessorPort {
    process(message: InboundMessage | null, update?: unknown): Promise<void>;
}

export const INBOUND_MESSAGE_PROCESSOR = Symbol("INBOUND_MESSAGE_PROCESSOR");

@Injectable()
export class JobSchedulerService implements OnModuleDestroy {
    private readonly logger = new Logger(JobSchedulerService.name);
    private readonly timers = new Set<NodeJS.Timeout>();

    public constructor(
        private readonly mediaGroupStore: MediaGroupBufferRepository,
        private readonly replyGenerationService: ReplyGenerationService,
        private readonly sessionService: SessionService,
        @Inject(forwardRef(() => CompactResultSenderService))
        private readonly compactResultSender: CompactResultSenderService,
        @Inject(INBOUND_MESSAGE_PROCESSOR)
        private readonly inboundMessageProcessor: InboundMessageProcessorPort
    ) {}

    public enqueueReplyGeneration(message: InboundMessage): void {
        void this.replyGenerationService.handle(message).catch((error: unknown) => {
            this.logger.error(`Reply generation failed update_id=${message.updateId}`, error as Error);
        });
    }

    public scheduleMediaGroupFlush(key: string, expectedDeadlineAt: number, waitDurationMs: number): void {
        const timer = setTimeout(() => {
            this.timers.delete(timer);
            void this.flushMediaGroup(key, expectedDeadlineAt);
        }, waitDurationMs);
        this.timers.add(timer);
    }

    public enqueueSessionCompact(chatId: string): void {
        void this.sessionService
            .compact(chatId)
            .then(result => this.compactResultSender.send(chatId, result))
            .catch((error: unknown) => {
                this.logger.error(`Session compact failed chat_id=${chatId}`, error as Error);
            });
    }

    public onModuleDestroy(): void {
        for (const timer of this.timers) {
            clearTimeout(timer);
        }
        this.timers.clear();
    }

    private async flushMediaGroup(key: string, expectedDeadlineAt: number): Promise<void> {
        const result = await this.mediaGroupStore.flush(key, expectedDeadlineAt);
        if (result.kind === "ready") {
            await this.inboundMessageProcessor.process(result.message);
        }
        if (result.kind === "pending") {
            this.scheduleMediaGroupFlush(key, expectedDeadlineAt, Math.round(result.waitDurationSeconds * 1000));
        }
    }
}
